Initial port of Cookie sharing needed for 4.5 apps to use new 5.0 DataProtection

HaoK opened the pull request:
Initial port of https://github.com/GrabYourPitchforks/aspnet5-samples/tree/dev/CookieSharing/Microsoft.Owin.Security.Cookies.Shareable

cc @blowdart @Tratcher @davidfowl

davidfowl commented on src/Microsoft.Owin.Security.Cookies.Shareable/AspNet5TicketDataFormat.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace Microsoft.Owin.Security.Cookies.Shareable
+{
+    public class AspNet5TicketDataFormat : SecureDataFormat<AuthenticationTicket>
```
Can we not call stuff AspNet5? :smile:

HaoK replied:
This is a weird one since this is basically a package that exists for old aspnet apps to target the AspNet5 data format...  Any ideas for a better name?

Tratcher replied:
You could just call it AspNet. Katana != AspNet.

HaoK replied:
AspNetTicketDataFormat?  InteropTicketDataFormat?  Or just nothing and let the namespace stand?

Tratcher replied:
If it's shared source then it just becomes TicketDataFormat

Tratcher commented on src/Microsoft.Owin.Security.Cookies.Shareable/AspNet5TicketDataFormat.cs:
```diff
@@ -0,0 +1,18 @@
+using Microsoft.Owin.Security.DataHandler.Encoder;
+using Microsoft.Owin.Security.DataProtection;
+
+namespace Microsoft.Owin.Security.Cookies.Shareable
```
Shareable->Interop?

Tratcher commented on src/Microsoft.Owin.Security.Cookies.Shareable/AspNet5TicketSerializer.cs:
```diff
@@ -0,0 +1,132 @@
+
+namespace Microsoft.Owin.Security.Cookies.Shareable
+{
+    public class AspNet5TicketSerializer : IDataSerializer<AuthenticationTicket>
```
Should this be shared source code to keep it in sync with the current version?

HaoK replied:
Sure good idea

PinpointTownes replied:
Though it's probably the main target for such a package, the cookies middleware is not the only one using a ticket serializer. Katana's OAuth2 bearer/server middleware also used it massively. Have you considering creating a more generic package for this class, to avoid tying it to cookies?

HaoK replied:
The name is a bit misleading.  The TicketSerializer should be reusable anywhere, the only thing that really interacts with cookies is via the sugar in UseCookieAuthentication.  We can certainly revisit how things are laid out/factored.

Tratcher commented on src/Microsoft.Owin.Security.Cookies.Shareable/DefaultCompatibilityConstants.cs:
```diff
@@ -0,0 +1,25 @@
+using System;
+using Microsoft.Owin.Security.Cookies;
+
+namespace Microsoft.AspNet.Identity
```
It's a little odd to have Identity in this package.

Tratcher replied:
It doesn't look like it's even used anywhere.

Tratcher commented on src/Microsoft.Owin.Security.Cookies.Shareable/CookieAuthenticationExtensions.cs:
```diff
@@ -0,0 +1,44 @@
+{
+    public static class CookieAuthenticationExtensions
+    {
+        public static IAppBuilder UseCookieAuthentication(
```
Should there be a sample for this scenario?

HaoK replied:
yeah I'll add a sample good idea

HaoK replied:
I think the full sample should live in identity repo since it pulls in a ton of stuff to actually do the full flow (EF + MVC + Identity cookie sharing).  But a tiny cookie sharing sample should be fine too

HaoK replied:
I'll file an issue for now, I'll do the intial end to end identity + everything sample first since that's what people actually will want to use

HaoK replied:
https://github.com/aspnet/Security/issues/412

HaoK commented:
Updated PR:
- Renamed package from Shareable => Microsoft.Owin.Security.Cookies.Interop
- Got rid of the overload / ShareableCookiesMiddleware gimmick that was just replacing the IDataProtectionProvider.  Instead added it to CookieAuthenticationOptions and if set, the middleware will use the options provider IGNORING the one from DI.

HaoK commented:
Sample will come in a follow up identity repo PR

Tratcher commented on src/Microsoft.Owin.Security.Cookies.Interop/AspNet5TicketDataFormat.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace Microsoft.Owin.Security.Cookies.Interop
+{
+    public class AspNet5TicketDataFormat : SecureDataFormat<AuthenticationTicket>
```
Drop the 5 from type names.

Tratcher commented on src/Microsoft.Owin.Security.Cookies.Interop/AspNet5TicketSerializer.cs:
```diff
@@ -0,0 +1,131 @@
+
+        private void Write(BinaryWriter writer, AuthenticationTicket model)
+        {
+            writer.Write(FormatVersion);
```
You'll need to account for @PinpointTownes's PR.

muratg commented:
@haok will this go in beta 8 or rc1?

HaoK commented:
Rc1 at this point. Unexpected complication that needs discussion

muratg commented:
@HaoK Will there be a new PR?  If so, should we close this?

HaoK commented:
I will just update this pr. I think we have a design now

HaoK commented:
@Tratcher updated the PR and added tests that verify both directions with a shared dataprotector
Interop cookie app => Aspnet5 cookie app
Aspnet 5 cookie => interop cookie app

The test code is not pretty as I had to use both versions of TestServer in the same file...ughhh

Tratcher commented on src/Microsoft.Owin.Security.Cookies.Interop/AspNetTicketSerializer.cs:
```diff
@@ -0,0 +1,221 @@
+            {
+                return null;
+            }
+
```
Needs a comment that says we will only read the first identity, any others will be ignored.

HaoK replied:
We have some wiggle room for how we want this to behave but for now that seems like a good default.  I'll update the comment

Tratcher commented:
Sample?

Tratcher commented:
:shipit:

HaoK commented:
Sample won't make Rc1, as it will prob fall out of the all up identity end to end flow.  But at least we have coverage of this now so we can check it in :)

HaoK commented:
e72a5639024eba48a021960f2bc141372c337afc
